README.md
# Flutter Spin Wheel Menu

[![pub package](https://img.shields.io/pub/v/flutter_zoom_drawer.svg)](https://pub.dev/packages/flutter_spin_wheel_menu) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

A Flutter package with custom implementation of the Card / Menu spinning like a wheel

## 🌟 Getting Started

To start using this package, add `flutter_spin_wheel_menu` dependency to your `pubspec.yaml`

```yaml
dependencies:
  flutter_spin_wheel_menu: "<latest_release>"
```

## 📌 Simple Example (Thanks to @ChidiebereEdeh)

```dart
import 'package:flutter/material.dart';
import 'package:flutter_spin_wheel_menu/flutter_spin_wheel_menu.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('FlutterSpinWheelMenu Example'),
      ),
      body: Center(
        child: FlutterSpinWheelMenu(
          rotationDuration: Duration(seconds: 10),
          radius: 80.0,
          perspective: 0.01,
          backgroundColor: Colors.blue,
          onTap: (index) {
            // Add your logic here
          },
          cardTitles: ['Item 1', 'Item 2', 'Item 3'],
          cardDescriptions: ['Description 1', 'Description 2', 'Description 3'],
          cardIcons: [Icons.star, Icons.favorite, Icons.star_border],
          numCards: 3,
        ),
      ),
    );
  }
}

```

## 📝 Documentation

```dart
    FlutterSpinWheelMenu(
        rotationDuration: Duration(seconds: 10),
        radius: 80.0,
        perspective: 0.01,
        backgroundColor: Colors.blue,
        onTap: (index) {
        print('Card clicked: $index');
          // Add your logic here
        },
        cardTitles: ['Item 1', 'Item 2', 'Item 3'],
        cardDescriptions: ['Description 1', 'Description 2', 'Description 3'],
        cardIcons: [Icons.star, Icons.favorite, Icons.star_border],
        numCards: 3,
    ),
```

## Screens

![Example app Image](https://drive.google.com/file/d/1JYM_rl9qmK_cuHMb8-ZKZy_Bm2egUjSm/view)

![Example app Demo](https://drive.google.com/file/d/1JZZJm-0KnGwxNWMdcm-gAn_irP_7piDQ/view)


## Issues

Please file any issues, bugs or feature request as an issue on our [GitHub](https://github.com/medyas/flutter_spin_wheel_menu/issues) page.

## Want to contribute

If you would like to contribute to the plugin (e.g. by improving the documentation, solving a bug or adding a cool new feature), please carefully review our [contribution guide](CONTRIBUTING.md) and send us your [pull request](https://github.com/fussion-integrator/flutter_spin_wheel_menu/pulls).

## Credits

Credits goes to [ChidiebereEdeh](https://github.com/fussion-integrator) as most of this package comes from his implementation.


CHANGELOG.md
[0.0.3] - Update on the ReadMe.
Updated the sample image and video links for the readme.

[0.0.2] - Update on the ReadMe.
The readme file was updated.

[0.0.1] - Initial release.
Initial package release.

test/flutter_spin_wheel_menu_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_spin_wheel_menu/flutter_spin_wheel_menu.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/flutter_spin_wheel_menu.dart
import 'dart:math';
import 'package:flutter/material.dart';

class FlutterSpinWheelMenu extends StatefulWidget {
  final Duration rotationDuration;
  final double radius;
  final double perspective;
  final Color backgroundColor;
  final Function(int) onTap;
  final List<String> cardTitles;
  final List<String> cardDescriptions;
  final List<IconData> cardIcons;
  final int numCards;

  const FlutterSpinWheelMenu({
    Key? key,
    this.rotationDuration = const Duration(seconds: 15),
    this.radius = 50.0,
    this.perspective = 0.01,
    this.backgroundColor = const Color(0xFF252A4D),
    required this.onTap,
    required this.cardTitles,
    required this.cardDescriptions,
    required this.cardIcons,
    required this.numCards,
  }) : super(key: key);

  @override
  _FlutterSpinWheelMenuState createState() => _FlutterSpinWheelMenuState();
}

class _FlutterSpinWheelMenuState extends State<FlutterSpinWheelMenu>
    with SingleTickerProviderStateMixin {
  late AnimationController _rotationController;
  double rotation = 0.0;

  @override
  void initState() {
    super.initState();
    _rotationController = AnimationController(
      vsync: this,
      duration: widget.rotationDuration,
    )..addListener(() {
      setState(() {
        rotation = _rotationController.value * 2 * pi;
      });
    });

    _rotationController.repeat();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onPanUpdate: (details) {
        setState(() {
          _rotationController.stop();
          rotation += details.delta.dx / 100;
        });
      },
      onPanEnd: (details) {
        _rotationController.repeat();
      },
      child: Stack(
        alignment: Alignment.center,
        children: List.generate(
          widget.numCards,
              (index) => _buildCircularPositionedWidget(index),
        ),
      ),
    );
  }

  Widget _buildCircularPositionedWidget(int index) {
    final double angle = (2 * pi / widget.numCards) * index;
    final double x = cos(angle + rotation) * widget.radius;
    final double y = sin(angle + rotation) * widget.radius;
    final double scale = 1 / (1 - y * widget.perspective);

    return Positioned(
      left: x + MediaQuery.of(context).size.width / 2 - 80,
      top: y + MediaQuery.of(context).size.height / 2 - 250,
      child: GestureDetector(
        onTap: () {
          widget.onTap(index);
        },
        child: Transform(
          transform: Matrix4.identity()
            ..setEntry(3, 2, widget.perspective)
            ..translate(x, y)
            ..scale(scale),
          child: Container(
            width: 120,
            height: 160,
            padding: const EdgeInsets.all(16),
            decoration: BoxDecoration(
              color: widget.backgroundColor,
              borderRadius: BorderRadius.circular(12),
            ),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                CircleAvatar(
                  radius: 30,
                  backgroundColor: Colors.white,
                  child: Icon(
                    widget.cardIcons[index],
                    size: 40,
                    color: widget.backgroundColor,
                  ),
                ),
                const SizedBox(height: 10),
                Text(
                  widget.cardTitles[index],
                  style: const TextStyle(
                    color: Colors.white,
                    fontWeight: FontWeight.bold,
                    fontSize: 14,
                  ),
                ),
                const SizedBox(height: 5),
                Text(
                  widget.cardDescriptions[index],
                  style: const TextStyle(
                    color: Colors.white,
                    fontSize: 10,
                  ),
                  textAlign: TextAlign.center,
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  @override
  void dispose() {
    _rotationController.dispose();
    super.dispose();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_spin_wheel_menu/flutter_spin_wheel_menu.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('FlutterSpinWheelMenu Example'),
      ),
      body: Center(
        child: FlutterSpinWheelMenu(
          rotationDuration: Duration(seconds: 10),
          radius: 80.0,
          perspective: 0.01,
          backgroundColor: Colors.blue,
          onTap: (index) {
            // Add your logic here
          },
          cardTitles: ['Item 1', 'Item 2', 'Item 3'],
          cardDescriptions: ['Description 1', 'Description 2', 'Description 3'],
          cardIcons: [Icons.star, Icons.favorite, Icons.star_border],
          numCards: 3,
        ),
      ),
    );
  }
}
